### ISSUE TITLE
 Cannot convert input parameter after converting to isolated worker model

### ISSUE THREAD

**Issue by Camios** [Azure/azure-functions-dotnet-worker#2444]
### Description

I had a .net 6 in-process model function app which has a HTTP Trigger and a CosmosDBInput binding.
I ran the .net upgrade assistant to convert to a .net 8 isolated worker model. It claimed success but left me with a bunch of Cosmos binding related errors. I tried fixing them manually and got it running, but now I'm getting a really strange runtime exception "FunctionInputConverterException".

Related complaint: The [examples section for isolated model](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-cosmosdb-v2-input?tabs=python-v2%2Cisolated-process%2Cnodejs-v4%2Cextensionv4&pivots=programming-language-csharp#example) if very sparse and deserves to updated to match the in-memory model

As far as I can tell from the Usage section, it should support IEnumerable and if I don't specify Id or SqlQuery, then it will return all documents.

```
[Function("Function1")]
public IActionResult Run
    [HttpTrigger(AuthorizationLevel.Function, "get")] 
    HttpRequest req,
    [CosmosDBInput(
        "EditorList", 
        "Items",
        Connection = "AccountEndpoint=https://localhost:8081/;AccountKey=<snipped>")]
     IEnumerable<object> documents)
```

The CosmosDBInput binding is just supposed to return all items in the database's container but fails with below exception.  I'm not using any variables from the route (which is a known bug)

```
2024-05-07T03:20:30.379Z] Function 'Function1', Invocation id '31e4bc2e-0605-4d15-8a64-dd0c48e06d91': An exception was thrown by the invocation.
[2024-05-07T03:20:30.382Z] Result: Function 'Function1', Invocation id '31e4bc2e-0605-4d15-8a64-dd0c48e06d91': An exception was thrown by the invocation.
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'Function1': Cannot convert input parameter 'documents' to type 'System.Collections.Generic.IEnumerable`1[[System.Object, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]' from type 'Microsoft.Azure.Functions.Worker.Grpc.Messages.GrpcModelBindingData'. Error:System.InvalidOperationException: Cosmos DB connection configuration 'AccountEndpoint=https://localhost:8081/;AccountKey=C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==' does not exist. Make sure that it is a defined App Setting.
[2024-05-07T03:20:30.384Z]    at Microsoft.Azure.Functions.Worker.CosmosDBBindingOptionsSetup.Configure(String connectionName, CosmosDBBindingOptions options) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\Config\CosmosDBBindingOptionsSetup.cs:line 37
[2024-05-07T03:20:30.387Z]    at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
[2024-05-07T03:20:30.389Z]    at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
[2024-05-07T03:20:30.391Z]    at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
[2024-05-07T03:20:30.394Z]    at System.Lazy`1.CreateValue()
[2024-05-07T03:20:30.396Z]    at Microsoft.Extensions.Options.OptionsCache`1.GetOrAdd[TArg](String name, Func`3 createOptions, TArg factoryArgument)
[2024-05-07T03:20:30.398Z]    at Microsoft.Extensions.Options.OptionsMonitor`1.Get(String name)
[2024-05-07T03:20:30.400Z]    at Microsoft.Azure.Functions.Worker.CosmosDBConverter.CreateCosmosClient[T](CosmosDBInputAttribute cosmosAttribute) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\CosmosDBConverter.cs:line 174
[2024-05-07T03:20:30.401Z]    at Microsoft.Azure.Functions.Worker.CosmosDBConverter.CreateTargetObjectAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\CosmosDBConverter.cs:line 91
[2024-05-07T03:20:30.403Z]    at Microsoft.Azure.Functions.Worker.CosmosDBConverter.ToTargetTypeAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\CosmosDBConverter.cs:line 85
[2024-05-07T03:20:30.405Z]    at Microsoft.Azure.Functions.Worker.CosmosDBConverter.ConvertFromBindingDataAsync(ConverterContext context, ModelBindingData modelBindingData) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\CosmosDBConverter.cs:line 57
[2024-05-07T03:20:30.407Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 94
[2024-05-07T03:20:30.409Z]    at IsolatedFunctionApps.DirectFunctionExecutor.ExecuteAsync(FunctionContext context) in C:\Users\camiller\source\repos\IsolatedFunctionApps\IsolatedFunctionApps\obj\Debug\net8.0\Microsoft.Azure.Functions.Worker.Sdk.Generators\Microsoft.Azure.Functions.Worker.Sdk.Generators.FunctionExecutorGenerator\GeneratedFunctionExecutor.g.cs:line 31
[2024-05-07T03:20:30.410Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
[2024-05-07T03:20:30.412Z]    at Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.FunctionsHttpProxyingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\extensions\Worker.Extensions.Http.AspNetCore\src\FunctionsMiddleware\FunctionsHttpProxyingMiddleware.cs:line 48
[2024-05-07T03:20:30.413Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 89
Stack:    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 94
[2024-05-07T03:20:30.415Z]    at IsolatedFunctionApps.DirectFunctionExecutor.ExecuteAsync(FunctionContext context) in C:\Users\camiller\source\repos\IsolatedFunctionApps\IsolatedFunctionApps\obj\Debug\net8.0\Microsoft.Azure.Functions.Worker.Sdk.Generators\Microsoft.Azure.Functions.Worker.Sdk.Generators.FunctionExecutorGenerator\GeneratedFunctionExecutor.g.cs:line 31
[2024-05-07T03:20:30.417Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
[2024-05-07T03:20:30.418Z]    at Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.FunctionsHttpProxyingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\extensions\Worker.Extensions.Http.AspNetCore\src\FunctionsMiddleware\FunctionsHttpProxyingMiddleware.cs:line 48
[2024-05-07T03:20:30.419Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 89.
[2024-05-07T03:20:30.584Z] Executed 'Functions.Function1' (Failed, Id=31e4bc2e-0605-4d15-8a64-dd0c48e06d91, Duration=1524ms)
[2024-05-07T03:20:30.586Z] System.Private.CoreLib: Exception while executing function: Functions.Function1. System.Private.CoreLib: Result: Failure
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'Function1': Cannot convert input parameter 'documents' to type 'System.Collections.Generic.IEnumerable`1[[System.Object, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]' from type 'Microsoft.Azure.Functions.Worker.Grpc.Messages.GrpcModelBindingData'. Error:System.InvalidOperationException: Cosmos DB connection configuration 'AccountEndpoint=https://localhost:8081/;AccountKey=C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==' does not exist. Make sure that it is a defined App Setting.
[2024-05-07T03:20:30.589Z]    at Microsoft.Azure.Functions.Worker.CosmosDBBindingOptionsSetup.Configure(String connectionName, CosmosDBBindingOptions options) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\Config\CosmosDBBindingOptionsSetup.cs:line 37
[2024-05-07T03:20:30.592Z]    at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
[2024-05-07T03:20:30.594Z]    at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
[2024-05-07T03:20:30.596Z]    at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
[2024-05-07T03:20:30.598Z]    at System.Lazy`1.CreateValue()
[2024-05-07T03:20:30.601Z]    at Microsoft.Extensions.Options.OptionsCache`1.GetOrAdd[TArg](String name, Func`3 createOptions, TArg factoryArgument)
[2024-05-07T03:20:30.603Z]    at Microsoft.Extensions.Options.OptionsMonitor`1.Get(String name)
[2024-05-07T03:20:30.605Z]    at Microsoft.Azure.Functions.Worker.CosmosDBConverter.CreateCosmosClient[T](CosmosDBInputAttribute cosmosAttribute) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\CosmosDBConverter.cs:line 174
[2024-05-07T03:20:30.607Z]    at Microsoft.Azure.Functions.Worker.CosmosDBConverter.CreateTargetObjectAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\CosmosDBConverter.cs:line 91
[2024-05-07T03:20:30.609Z]    at Microsoft.Azure.Functions.Worker.CosmosDBConverter.ToTargetTypeAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\CosmosDBConverter.cs:line 85
[2024-05-07T03:20:30.612Z]    at Microsoft.Azure.Functions.Worker.CosmosDBConverter.ConvertFromBindingDataAsync(ConverterContext context, ModelBindingData modelBindingData) in D:\a\_work\1\s\extensions\Worker.Extensions.CosmosDB\src\CosmosDBConverter.cs:line 57
[2024-05-07T03:20:30.614Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 94
[2024-05-07T03:20:30.616Z]    at IsolatedFunctionApps.DirectFunctionExecutor.ExecuteAsync(FunctionContext context) in C:\Users\camiller\source\repos\IsolatedFunctionApps\IsolatedFunctionApps\obj\Debug\net8.0\Microsoft.Azure.Functions.Worker.Sdk.Generators\Microsoft.Azure.Functions.Worker.Sdk.Generators.FunctionExecutorGenerator\GeneratedFunctionExecutor.g.cs:line 31
[2024-05-07T03:20:30.618Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
[2024-05-07T03:20:30.620Z]    at Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.FunctionsHttpProxyingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\extensions\Worker.Extensions.Http.AspNetCore\src\FunctionsMiddleware\FunctionsHttpProxyingMiddleware.cs:line 48
[2024-05-07T03:20:30.623Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 89
[2024-05-07T03:20:30.625Z]    at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request) in D:\a\_work\1\s\src\DotNetWorker.Grpc\Handlers\InvocationHandler.cs:line 88
Stack:    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 94
[2024-05-07T03:20:30.626Z]    at IsolatedFunctionApps.DirectFunctionExecutor.ExecuteAsync(FunctionContext context) in C:\Users\camiller\source\repos\IsolatedFunctionApps\IsolatedFunctionApps\obj\Debug\net8.0\Microsoft.Azure.Functions.Worker.Sdk.Generators\Microsoft.Azure.Functions.Worker.Sdk.Generators.FunctionExecutorGenerator\GeneratedFunctionExecutor.g.cs:line 31
[2024-05-07T03:20:30.628Z]    at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
[2024-05-07T03:20:30.629Z]    at Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.FunctionsHttpProxyingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\extensions\Worker.Extensions.Http.AspNetCore\src\FunctionsMiddleware\FunctionsHttpProxyingMiddleware.cs:line 48
[2024-05-07T03:20:30.631Z]    at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 89
[2024-05-07T03:20:30.633Z]    at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request) in D:\a\_work\1\s\src\DotNetWorker.Grpc\Handlers\InvocationHandler.cs:line 88.
```

### Steps to reproduce

1. Install and run Azure Cosmos DB Emulator. Wait a long time for the web page to load the UI. Note the primary connection string.
2. Extract [IsolatedFunctionApps.zip](https://github.com/Azure/azure-functions-dotnet-worker/files/15228998/IsolatedFunctionApps.zip)
3. Open the solution in Visual Studio and modify Connection string in the CosmosDBInput binding if necessary
4. Start local debug session
5. Note the console window shows the address for the function app's http trigger
6. use postman to send request to the url of the function app
7. Get exception

**Comment by Camios** [comment]
I tried generating a new solution for minimal reproduction, but I don't get the same problem.

**Comment by liliankasem** [comment]
Just to clarify, a new solution with this function is working just fine as expected?

It seems like the migration tool struggled here, so maybe there are improvements to make to that. However without a repro, it will be difficult to investigate this. It could be an issue with the project file, and/or nuget references - are you on all the latest packages for the worker, worker sdk, and cosmos extension?

Re: samples - we have more scenarios documented here: https://github.com/Azure/azure-functions-dotnet-worker/blob/main/samples/Extensions/CosmosDB/CosmosInputBindingFunctions.cs

**Comment by Camios** [comment]
@liliankasem 

> Just to clarify, a new solution with this function is working just fine as expected?

Yes


I found the upgrade assistant failed in the many ways but it was its oversight of not adding one or both of these packages to the project which is causing the runtime exception about missing file `Microsoft.Bcl.AsyncInterfaces`:
    - `Microsoft.ApplicationInsights.WorkerService`
    - `Microsoft.Azure.Functions.Worker.ApplicationInsights`

Here's a longer list of things the upgrade assistant didn't do to my existing Azure Function App project (mix of problems it left behind and things it doesn't do compared to new project):
- Didn't add package dependencies:
    - `Microsoft.Azure.Functions.Worker.Extensions.CosmosDB` 
    - `Microsoft.ApplicationInsights.WorkerService`
    - `Microsoft.Azure.Functions.Worker.ApplicationInsights`
- Didn't run a package restore after changing from in-process worker model based dependencies (WebJobs) to isolated worker model dependencies.
- _Possibly_ didn't trigger an Azure core tools update?? Might be a once-off.
- Didn't convert `CosmosDB` parameter-based attribute output bindings (IAsyncCollector) to method-based `CosmosDBOutput` attributes and doesn't convert the parameter to a return value. 
- Didn't rename named binding parameters like `collectionName` and `ConnectionStringSetting`.
- Didn't convert System.Web.Http usage:
    - new QueryCollection should have been converted to QueryString.Create
    - HttpRequest no longer has a ReadAsStringAsync, needs to be converted to create `StreamReader` from `HttpRequest.Body`
    - Was using `<Something>MessageResult` and they needed to be converted to `BadRequestObjectResult`, `StatusCodeResult`, `ObjectResults`, etc
- Had Dependency injection in `startup.cs` with `[assembly: FunctionsStartup(typeof(Startup))]` but I think it should have converted to `HostBuilder.ConfigureServices` in the newly added `prrogram.cs` and removed `startup.cs`?
- In csproj it added `<ImplicitUsings>enabled</ImplicitUsings>`, but pretty sure `"enabled"` should be `"enable"` (or `"true"`). 
- It went from using `Newtonsoft.Json` serializers to `System.Text.Json` but didn't convert any of the POCO models that were using Newtonsoft-specific attributes. This annoyingly causes runtime exceptions. And there's [no way to swap back to Newtonsoft](https://github.com/Azure/azure-functions-dotnet-worker/issues/2131)  and `System.Text.Json` has functional differences which might be breaking (like a lack of property-level `JsonProperty` `DefaultValue` handling)
- New project created a mix of files in the Properties folder

It should be doing best effort to convert and warning if there's a problem.

**Comment by liliankasem** [comment]
Thank you for the detailed write up! I'm going to consider the original issue resolved and repurpose this issue to track migration assistant enhancements - sound fair?

**Comment by Camios** [comment]
From my perspective:

1.  I'm not blocked by the runtime exception anymore and the focus turns more to addressing improvements to the upgrade assistant.
2. I did have some outstanding questions (I bolded them in my previous command) and I was hoping you could answer promptly to help my migration. 




>Re: samples - we have more scenarios documented here: https://github.com/Azure/azure-functions-dotnet-worker/blob/main/samples/Extensions/CosmosDB/CosmosInputBindingFunctions.cs

Thanks for the link, I just found the Examples on learn.microsoft.com were insufficient and ideally expanded. And if it doesn't exist then linking to the github samples from learn.microsoft.com page would be helpful too.

**Comment by liliankasem** [comment]
> Had Dependency injection in startup.cs with [assembly: FunctionsStartup(typeof(Startup))] but I think it should have converted to HostBuilder.ConfigureServices in the newly added prrogram.cs and removed startup.cs?

Correct, DI should be moved to Program.cs and Startup.cs should be removed.

> It went from using Newtonsoft.Json serializers to System.Text.Json but didn't convert any of the POCO models that were using Newtonsoft-specific attributes. This annoyingly causes runtime exceptions. And there's https://github.com/Azure/azure-functions-dotnet-worker/issues/2131 and System.Text.Json has functional differences which might be breaking (like a lack of property-level JsonProperty DefaultValue handling)

You should be able to configure a Newtonsoft serializer ([example](https://github.com/Azure/azure-functions-dotnet-worker/blob/b0e717808a26099b3eaac8e7cf723064dfbb727a/samples/Configuration/Program.cs#L58-L69)). There is unfortunately a bug with this when using AspNetCore `.ConfigureFunctionsWebApplication`. But you should still be able to do this using `.ConfigureFunctionsWorkerDefaults` if you don't need AspNetCore. The Cosmos extension uses the worker config serializer (as shown in the example linked) as the serializer for the CosmosClientOptions.

> New project created a mix of files in the Properties folder, but they weren't added during the upgrade to the existing project. Does it need them?

I think you only need `launchSettings.json` for VS, the rest can probably be removed.

> The new project also added <FrameworkReference Include="Microsoft.AspNetCore.App" /> which wasn't in the existing project wasn't added to the upgraded project. Should the upgrade assistant be adding AspNetCore framework reference to upgraded Function Apps too?

`<FrameworkReference Include="Microsoft.AspNetCore.App" />` is only required if you want to use the new HTTP extension `Worker.Extensions.Http.AspNetCore` (same for `ConfigureFunctionsWebApplication`), otherwise you don't need it.

**Comment by Camios** [comment]
>> It went from using Newtonsoft.Json serializers to System.Text.Json but didn't convert any of the POCO models that were using Newtonsoft-specific attributes. This annoyingly causes runtime exceptions. And there's https://github.com/Azure/azure-functions-dotnet-worker/issues/2131 and System.Text.Json has functional differences which might be breaking (like a lack of property-level JsonProperty DefaultValue handling)

> You should be able to configure a Newtonsoft serializer ([example](https://github.com/Azure/azure-functions-dotnet-worker/blob/b0e717808a26099b3eaac8e7cf723064dfbb727a/samples/Configuration/Program.cs#L58-L69)). There is unfortunately a bug with this when using AspNetCore  `.ConfigureFunctionsWebApplication`. But you should still be able to do this using `.ConfigureFunctionsWorkerDefaults` if you don't need AspNetCore. The Cosmos extension uses the worker config serializer (as shown in the example linked) as the serializer for the CosmosClientOptions.

When debugging the Cosmos input binding's JSON deserializing, I saw `Microsoft.Azure.Functions.Worker.CosmosDBConverter` has a hard-coded `JsonSerializerOptions` when deserializing which would prevent any settings override in Program.cs. 

Is that `CosmosDBConverter` only used by Function Apps with the ASP.NET Core integrations (is that the bug you refer to)?
Is a different DB converter used when the Function app doesn't use ASP.NET Core integration?

**Comment by liliankasem** [comment]
> CosmosDBConverter has a hard-coded JsonSerializerOptions when deserializing which would prevent any settings override in Program.cs.

Yes it does have a hardcoded `JsonSerializerOptions`; what I was trying to convey is that there is a difference between that serializer and the cosmos client serializer that _can_ be configured. The one that is hardcoded in the Converter is used when binding to a [POCO](https://github.com/Azure/azure-functions-dotnet-worker/blob/1053f8bb6cbe88a2976bdff2480a3cb24ff1dd97/extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs#L116). And the other is used by the [CosmoClient](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.cosmos.cosmosclientoptions.serializer?view=azure-dotnet), this one can be configured via workerOptions.Serializer.

Wanting the latter one to be configurable is perfectly valid feedback, and likely was an oversight. Please open a new issue to help us triage and address that.

> Is that CosmosDBConverter only used by Function Apps with the ASP.NET Core integrations (is that the bug you refer to)?
Is a different DB converter used when the Function app doesn't use ASP.NET Core integration?

No, this converter is used for all Cosmos **input** bindings.
